### PR TITLE
Allow customizing Compose compiler plugin

### PR DESCRIPTION
**Pull request by AlexeyTsvetkov**
By default, the Compose Multiplatform Gradle plugin uses `org.jetbrains.compose.compiler:compiler:<COMPOSE_VERSION>` as a compiler plugin .
However, a new version of Kotlin might be incompatible with the default version of compiler plugin. Previously, that forced users to update to a new version of Compose Gradle plugin & Compose libraries in order to use a new version of Kotlin. Accordingly, Compose framework developers had to release a new version of all libraries, when a new version of Kotlin is released.

Some time ago the Compose team at Google made it possible to update a compiler plugin without updating the Gradle plugin and/or all Compose libraries https://developer.android.com/jetpack/androidx/releases/compose-kotlin

This change allows to specify a custom compiler artifact by using the following DSL:
```
compose {
    kotlinCompilerPlugin.set("<VERSION<")
    // or
    kotlinCompilerPlugin.set("<GROUP_ID>:<ARTIFACT_ID>:<VERSION>")
}
```

### CODE REVIEW

**Comment by igordmn** [comment]
>  kotlinCompilerPlugin.set("<VERSION<")

Can we use a property? It is more idiomatic for Gradle Kotlin DSL:
```
kotlinCompilerPlugin = "1.3.1"
```

**Comment by AlexeyTsvetkov** [comment]
@igordmn I'm not sure. Gradle [seems to be moving](https://github.com/orgs/gradle/projects/31/views/1) in direction of switching to Provider API everywhere in Gradle 9.  Also they are [going](https://github.com/gradle/build-tool-roadmap/issues/38) to add an assignment operator overload to `Property` type, so soon users will be able to write `kotlinCompilerPlugin = "1.3.1"`, while retaining laziness of Provider API

**Comment by igordmn** [comment]
> kotlinCompilerPlugin.set("<GROUP_ID>:<ARTIFACT_ID>:<VERSION>")

We should also prepare a page with tutorial how to to that in with mapping table. Also, worth to add our own exception in Gradle plugin with a link to that page.

**Comment by igordmn** [comment]
> @igordmn I'm not sure. 

Ok, let's keep the current way

**Review by dima-avdeev-jb — APPROVED** [review]

File: gradle-plugins/compose/build.gradle.kts [hunk 1]
```diff
@@ -177,8 +178,13 @@ tasks.withType<Test>().configureEach {
...
     dependsOn(":publishToMavenLocal")
-    systemProperty("compose.plugin.version", BuildProperties.deployVersion(project))
-    systemProperty("kotlin.version", project.property("kotlin.version").toString())
+
+    systemProperty("compose.tests.compose.gradle.plugin.version", BuildProperties.deployVersion(project))
+    for ((k, v) in project.properties) {
+        if (k.startsWith("compose.")) {
+            systemProperty(k, v.toString())
+        }
+    }
```
[the remark this answers is not in the document]
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/build.gradle.kts:187]: We duplicate "compose.*" properties to tests system properties, so we can have easy access to the majority of relevant properties in tests

File: gradle-plugins/compose/build.gradle.kts [hunk 2]
```diff
@@ -177,8 +178,13 @@ tasks.withType<Test>().configureEach {
     configureJavaForComposeTest()
 
     dependsOn(":publishToMavenLocal")
-    systemProperty("compose.plugin.version", BuildProperties.deployVersion(project))
-    systemProperty("kotlin.version", project.property("kotlin.version").toString())
```
[the remark this answers is not in the document]
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/build.gradle.kts:181]: I think `kotlin.version` is misleading in this context. In reality we want `compose.compiler.compatible.kotlin.version`

**Pull request merged**